labs/CryptoProofs/CryptoProofs.md

```diff
@@ -0,0 +1,232 @@
+
+```
+matched_pt = join "tootough"
+matched_ct = 0x95d07f8a72707733
+```
+
+To make this solvable, try it again with the first six bytes of key provided: `0x1234567890ab`.
+
+## 2.2 Proof of Inversion
+
+For symmetric ciphers, it is necessary that the decrypt function *inverts* the encrypt function. (That is, it restores the ciphertext to the original plaintext.) It is easy to express this property in Cryptol.
+
+Consider the following functions `f` and `g`:
+
+```
+f: Integer -> Integer
+f x = 3 * x + 2
+g: Integer -> Integer
+g x = (x - 2) / 3
+```
+
+We want to prove that function `g` inverts function `f`; that is, applying `g` to the result of `f x` gets `x` back. Here's the invocation:
+
+```bash
+labs::CryptoProofs::CryptoProofs> :prove \x -> g (f x) == x
+Q.E.D.
+(Total Elapsed Time: 0.023s, using "Z3")
+```
+
+Here's the breadown of this proof:
+
+|Proof of Inversion||||
+|-|-|-|-|
+| `:prove`            | `\x`                   | `->`        | `g (f x) == x` |
+| "Prove to me" | "that for all `x`" | "it is true that" | "`g` inverts `f`" |
+|
+
+### Exercise 2.2.1 The other direction
+
+Our example proof showed that `g` inverts `f` for all inputs. Does this work the other way around? Try it! If the proof fails, it will provide a *counterexample*. Use the counterexample to understand what went wrong.
+
+### Exercise 2.2.2 DES inversion
+
+Use Cryptol to prove that `DES.encrypt` and `DES.decrypt` are inverses for all possible inputs. Show both directions.
+
+*Hint*: Note that this function has more than one input. To make it work, you can either use a tuple `(,)` (tuple-style) or a chain of of `\->` (Curried-style) constructs for your lambda function input.
+
+*Hint*: For fastest results, use the `abc` prover.
+
+## 2.3 Collision Detection
+
+In cryptography, a *collision* occurs when two different inputs produce the same output. For some cryptographic functions, such as pseudo-random number generators (PRNGs), it may be desirable to demonstrate an absence of collisions. In other functions, such as cryptographic hash functions, collisions are inevitable, but should be difficult to discover. It is easy in Cryptol to ask the solver to search for collisions. (Though finding a solution may not be possible.)
+
+### Exercise 2.3.1 DES Key Collisions
+
+Use the solver to find two different keys and a plaintext such that both keys encrypt that plaintext to the same ciphertext.
+
+```bash
+labs::CryptoProofs::CryptoProofs> :s prover=yices
+labs::CryptoProofs::CryptoProofs> :sat \(k1,k2,pt) -> k1 != k2 /\ DES.encrypt k1 pt == DES.encrypt k2 pt
+(\(k1, k2, pt) -> k1 != k2 /\ DES.encrypt k1 pt == DES.encrypt k2
+                                                               pt)
+  (0x0000000000000000, 0x0100000000000000, 0x0000000000000000) = True
+(Total Elapsed Time: 1.662s, using Yices)
+```
```

Comment:
I think you should explain where this collision comes from (that the low bit of every byte of key is used as a checksum and not used during encryption).

We used to do a theorem showing that these extra bits could take on _any_ value. It looks something iike:
```
labs::CryptoProofs::CryptoProofs> :prove \(k:[56]) -> \(c1 : [8]) -> \(c2 : [8]) -> \pt -> DES.encrypt (rejigger k c1) pt == DES.encrypt (rejigger k c2) pt \
where rejigger a b = join [ ai#[bi] | ai <- split`{8} a | bi <- b ] 
Q.E.D.
(Total Elapsed Time: 0.473s, using "Yices")
```

Reply:
Ah, I see you do this below. Great!!!